ChopChop/Views/Graph/EditorGraphView.swift

```diff
@@ -0,0 +1,148 @@
+import SwiftUI
+
+struct EditorGraphView: View {
+    @ObservedObject var viewModel: EditorGraphViewModel
+    @ObservedObject var selection = SelectionHandler<RecipeStepNode>()
+
+    @GestureState var portalDragOffset = CGVector.zero
+    @GestureState var nodeDragOffset: EditorGraphViewModel.NodeDragInfo?
+    @GestureState var lineDragInfo: EditorGraphViewModel.LineDragInfo?
+    @GestureState var placeholderNodePosition: CGPoint?
+
+    var body: some View {
+        ZStack {
+            linesView
+
+            if let info = lineDragInfo, viewModel.isEditable {
+                placeholderLineView(info: info)
+            }
+
+            nodesView(nodes: viewModel.graph.topologicallySortedNodes)
+
+            if let position = placeholderNodePosition, viewModel.isEditable {
+                placeholderNodeView(position: position)
+            }
+        }
+        .frame(minWidth: 0, maxWidth: .infinity, minHeight: 0, maxHeight: .infinity)
+        .contentShape(Rectangle())
+        .clipped()
+        .navigationBarTitleDisplayMode(.inline)
+        .onTapGesture {
+            withAnimation {
+                selection.deselectAllNodes()
+            }
+        }
+        .gesture(
+            LongPressGesture()
+                .sequenced(before: DragGesture(minimumDistance: 0)
+                    .updating($placeholderNodePosition) { value, state, _ in
+                        state = value.location
+                    }
+                    .onEnded(viewModel.onLongPressPortal))
+                .exclusively(before: DragGesture()
+                    .updating($portalDragOffset) { value, state, _ in
+                        state = CGVector(dx: value.translation.width, dy: value.translation.height)
+                    }
+                    .onEnded(viewModel.onDragPortal)
+                )
+        )
+    }
+
+    var linesView: some View {
+        ForEach(viewModel.graph.edges, id: \.self) { edge in
+            if let offset = nodeDragOffset {
+                Line(from: (edge.source.position ?? .zero)
```

Comment:
is it possible for these calculations to be done in the ViewModel or Model?

Reply:
That's possible, but the ViewModel would then return a View, which seems odd.